tests/integration-all/api-gateway-websocket/tests.js

```diff
@@ -0,0 +1,104 @@
+'use strict';
+
+const path = require('path');
+const AWS = require('aws-sdk');
+const _ = require('lodash');
+const { expect } = require('chai');
+
+const { getTmpDirPath, readYamlFile, writeYamlFile } = require('../../utils/fs');
+const {
+  region,
+  // confirmCloudWatchLogs,
+  createTestService,
+  deployService,
+  removeService,
+} = require('../../utils/misc');
+const {
+  createApi,
+  deleteApi,
+  getRoutes,
+  createStage,
+  deleteStage,
+} = require('../../utils/api-gateway-v2');
+
+const CF = new AWS.CloudFormation({ region });
+
+describe('AWS - API Gateway Websocket Integration Test', function() {
+  this.timeout(1000 * 60 * 10); // Involves time-taking deploys
+  let serviceName;
+  let stackName;
+  let tmpDirPath;
+  let serverlessFilePath;
+  let websocketApiId;
+  const stage = 'dev';
+
+  before(() => {
+    tmpDirPath = getTmpDirPath();
+    console.info(`Temporary path: ${tmpDirPath}`);
+    serverlessFilePath = path.join(tmpDirPath, 'serverless.yml');
+    const serverlessConfig = createTestService(tmpDirPath, {
+      templateDir: path.join(__dirname, 'service'),
+    });
+    serviceName = serverlessConfig.service;
+    stackName = `${serviceName}-${stage}`;
+    console.info(`Deploying "${stackName}" service...`);
+    deployService(tmpDirPath);
+    // create an external websocket API
+    const externalWebsocketApiName = `${stage}-${serviceName}-ext-api`;
+    return createApi(externalWebsocketApiName).then(wsApiMeta => {
+      websocketApiId = wsApiMeta.ApiId;
+      return createStage(websocketApiId, 'dev');
+    });
+  });
+
+  after(() => {
+    // NOTE: deleting the references to the old, external websocket API
+    const serverless = readYamlFile(serverlessFilePath);
+    delete serverless.provider.apiGateway.websocketApiId;
+    writeYamlFile(serverlessFilePath, serverless);
+    // NOTE: we need to delete the stage before deleting the stack
+    // otherwise CF will refuse to delete the deployment because a stage refers to that
+    return deleteStage(websocketApiId, 'dev').then(() => {
+      // NOTE: deploying once again to get the stack into the original state
+      console.info('Redeploying service...');
+      deployService(tmpDirPath);
+      console.info('Removing service...');
+      removeService(tmpDirPath);
+      console.info('Deleting external rest API...');
+      return deleteApi(websocketApiId);
+    });
+  });
+
+  describe('Minimal Setup', () => {
+    it('should expose an accessible websocket endpoint', () =>
+      CF.describeStacks({ StackName: stackName })
```

Comment:
It'll be nice to setup some Websockets connection instead, and test that way, whether we get expected response.

Not sure though, how straightforward that is to do.

Reply:
I tried it before, but it is not straightforward.
you need to create a websocket, add a listener to its events, in the connection open event add listeners to the data received events.
also the whole setup is not a promise, so you need to add the `done` callback to the test (and if you not call it, the test will timeout after 10 minutes -- value taken from the REST integration but it says it's needed because of the CF deploy).

also, until you support integration response route on api gateway, the websocket will not return any data unless you explicitly `postToConnection`, requiring additional setup (and permissions) inside the lambda

Reply:
Ok, I see. So technically all that we can test, is that we've connected to given socket without issues

And eventual arrival of our message could eventually be tested via logging it, and inspecting the CloudWatch logs.

Reply:
I decided to give some hand on this, I looked into it, and updated tests to issue actual WebSockets connection and confirm afterwards that some CloudWatch logs are getting written for the server

I could also check the logs of lambda to confirm messages I've send, but I skipped on that part (anyway I confirmed manually that this does happen)